REVIEW 1 (reviews closed) - ADD: Documentation section to our packaging guide (#14)
pyOpenSci/python-package-guide: documentation/package-documentation-sphinx.md

> # Python Package Documentation

Author: After setting this up myself for the first time in a while i'm reminded of HOW confusing this ecosystem is. I think that this packaging guide is not enough. And we need some complementary tutorials to go along with it. These tutorials need to explain what extensions to use and how to set things up.

An example conf.py file in a repo with some documentation is probably ideal with tutorials that explain all of the parts. otherwise it's a huge rabbit hole.

Author: moved this over to slack as you saw :) i love looking at simpler options. it seems like most are using sphinx right now so we might have to start there to be consistent across the ecosystem and watch to see if things change. it looks like a really nice platform and if it's rooted in MD it's probably simpler to config from the start. i think most of the sphinx extensions are ones that kind of work around .rst.

* **API  documentation:** this is documentation contained within the docstrings of all user-facing functions, methods and classes in your package.  

## User-facing documentation for your Python package: 

User-facing documentation is often an easy to navigate website written with 
user friendly, less technical language. User facing documentation should be:

* easy to read (consider writing for a 12th grade audience)
* contain tutorials or vignettes that support getting started using your package 

### What tools to use in building your documentation: sphinx

Most python packages maintainers use [sphinx](https://www.sphinx-doc.org/) to build their documentation. 
Sphinx has documentation themes that give your docuemntation a clean online look that are easy to apply to your 
package's documentation. 

While you are free to use whatever theme works for you,
the most common modern templates that we see in the scientific 
community currently and recommend are: 

* [pydata-sphinx-theme](https://pydata-sphinx-theme.readthedocs.io/) 
* [sphinx-book-theme](https://sphinx-book-theme.readthedocs.io/)
* [furo](https://pradyunsg.me/furo/quickstart/)

> Both of the above themes support [myst](https://myst-parser.readthedocs.io/) which allows you to build your package 

Reviewer: Wow, I've learned a new useful thing. Thank you! :)

Author: yahoo!!

> documentation in `markdown` rather than `.rst`. 

Reviewer: Maybe it will be nice to add one sentence more, for example:

---

Why `markdown` instead of `restructured text`? There are three reasons. Most of your core text files within a project are in `.md` format, and GitHub / Jupiter Notebook styling uses this format.

Author: excellent suggest! fixed!

* [Here is our documentation if you want to see a myST example of a landing page.](https://github.com/pyOpenSci/python-package-guide/blob/main/index.md)
```

In addition to a well designed [README FILE](readme-file-best-practices), and a  
[contributing file](contributing-file), 
there are several core components of Python package documentation, 
including:

* **User-facing documentation:** This refers to easy-to-read documentation that helps a user work with your package. This documentation should help users both install and use the functionality of your package.  

> * **API  documentation:** API documentation is generated from [docstrings](https://pandas.pydata.org/docs/development/contributing_docstring.html) found in your code. Ideally you have docstrings for all user-facing functions, methods and classes in your Python package.  

Reviewer 2: Probably we should define what API stands for and explain how that relates to docstrings, i.e. the docstrings are closest to the API itself so it makes sense to that a significant portion of the API documentation can be generated from the in an automated fashion

Author: Great idea. also added how it's important for user facing docs ie `help()` !

## User-facing documentation for your Python package: 

User-facing documentation should be published on a easy-to-navigate website. All language should be written with non-developer users in mind. This means 
using language that is less technical.

A few tips to make sure your documentation is accessible include: 

* Whenever possible, define technical terms and jargon.

> * Consider writing writing instructions for a 12th grade level reader. 

Author: totally. not commit this only because i made both edits locally. thanks! fixed!

* Include step by step code examples, tutorials or vignettes that support getting started using your package.

> ### What tools to use to build your documentation: sphinx

Author: ok one thing to consider here is that 
1. we don't want to provide too many tools as examples. especially if the community is largely using sphinx. 
2.  we will tutorials that cover the specifics of all of this. this is a guide for what tools people should use. 
Finally myst is a plugin for sphinx so it's really different than (tell me if i'm wrong) mkdocs is a different engine altogether right? my understanding is it's a static site generator kind of like jekyll. so it's a whole different animal / engine than sphinx

> ### What tools to use to build your documentation: sphinx

Author: if a student was new only knew about markdown (i think about my students) and markdown worked, probably it would be cognitive overload to teach both rst and md. or i wouldn't chose to teach both! 

i think a simple breakout mentioning rst is warranted but i wouldn't go into detail in a guide as it risks serious cognitive overload in an already complicated ecosystem!

Most python packages use [sphinx](https://www.sphinx-doc.org/) to build their documentation. 

> Sphinx has documentation themes that are easy to apply and help you quickly and easily build an easy-to navigate website.

Author: ok! fixed locally!

Sphinx also offers extensions that support:

* [automatically documenting your API using docstrings in your code](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)
* [running and testing code examples in your docstrings (doctest)](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html)

While you are free to use whatever sphinx theme you prefer,
the most common modern templates that recommend for the Python scientific 
community are:  

* [pydata-sphinx-theme](https://pydata-sphinx-theme.readthedocs.io/) 
* [sphinx-book-theme](https://sphinx-book-theme.readthedocs.io/)
* [furo](https://pradyunsg.me/furo/quickstart/)

> All of the above themes support [myst](https://myst-parser.readthedocs.io/) 

Author: reworded the entire section!

which allows you to build your package documentation in using `markdown` 
rather than `.rst`. More on that below.

```{tip}
This book is created using sphinx and the `furo` theme.
```

### myST vs rst syntax to create your docs 
There are two commonly used syntaxes for creating docs:

> 1. [myST:](https://myst-parser.readthedocs.io/en/latest/intro.html) myST is a fusion of markdown and rST. It is a nice option if you are more comfortable writing markdown. But more flexible than markdown. 

Author: i'm pretty sure the second element here was rst. but i actually rewrote this entire thing to include vanilla markdown as well. frankly that was my first step into sphinx rst + md to support some md files (like readme).

2. You can host your documentation using [Read the Docs](https://readthedocs.org/).

If you don't want to maintain a documentation website for your Python package, 
we suggest using Read the Docs. Read the Docs allows you to:

* quickly launch a website using the documentation found in your GitHub repository.  
* track versions of your documentation as you release updates

## Documentation landing page best practices 

To make it easy for users to find what they need quickly, all packages should 
have 4 elements on the home page of their documentation:

* **Getting started:** This section should provide the user with a quick start for installing your package. A small example of how to use the package is good to have here as well. 
* **About:** Describe your project, state project goals and what it does. 
* **Community:** Instructions for ho to help and/or get involved. This section might include a development guide. 
* **Documentation:** The actual project documentation. You may have two sections of documentation - the user facing documentation and your API reference. 

> 

Author: this is the get started landing page. i'll add a screen shot of what this looks like. it's a basic landing page approach that everyone i think should use.  we use it on all of our guides as well. so there is no pip install they are literally all cards with links to key areas. it addresses the issue if guiding a user  where to go... 
<img width="906" alt="Screen Shot 2022-12-27 at 3 46 48 PM" src="https://user-images.githubusercontent.com/7649194/209731663-0a2baed1-7a2a-4c44-bd9a-c60c4adcbe6b.png">

> 

Author: making a note to add a screenshot for what this looks like to the guide. in our tutorials we will show them how to create the cards.

* Explains what every input and output variable's (type) is (ie. `string`, `int`, `np.array`)
* Explains the expected output `return` of the object, method or function.

> ### Python docstring format best-practices 

Reviewer 2: ```suggestion
### Python docstring format best practices 
```
Above you do not hyphenate best practices. Changing here for consistency.
Not sure a hyphen is needed here and if possible I would reword to avoid the noun stack, maybe just "Python docstring formats"?

There are several Python docstring formats that you can chose to use when documenting 

> your package. We suggest using [numpy-style docstrings](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard).

Reviewer 2: We should probably list the other two common formats and link to them as well.  
Then we should give a rationale for why we suggest using NumPy.  
I'm guessing the rationale is something like "because most other scientific packages adopt the NumPy format".
I also suggest writing "NumPy format" as the Napoleon documentation does, with NumPy capitalized as the developers write it, and without a hyphen:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html

There is also a link to the Google docstring format spec in the Napoleon index (https://google.github.io/styleguide/pyguide.html#Comments)

Author: i'll add it in a breakout. NOTE i'm intentionally trying to not list everything. i just think it's additional weight that a reader doesn't need especially if they are a beginner. it would be like my teaching my earth analytics course and trying to teach both r and python - both can do the same thing but they are different. there are a few other formats out there as well.

```{tip}

> If you are using `numpy-style` docstrings, be sure to include the sphinx napoleon 

Reviewer 2: ```suggestion
If you are using NumPY format docstrings, be sure to include the sphinx napoleon 
```

change `numpy-style` -> NumPy format (remove monospace or be consistent with monospace throughout)

Author: oh now i see where your comment came from. my goodness GitHub really did some weird things today with all of these comments. i just saw this @arianesasso by itself in a list of other comments. i've made this change throughout  except where do you see the Y capitalized? https://numpy.org/ i don't see that in the numpy docs :) i almost wrote dogs :)

extension in your documentation `conf.py` file. This extension allows sphinx 
to properly read and format numpy-style docstrings. 
```

### Docstring examples Better and Best 

> 

Reviewer 2: Is this copied verbatim from geopandas?
I think there's usually line breaks between the short description, Parameters, and Returns.
And it should be "Returns" with an "s", not "Return".
See example here:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy

```Python
def extent_to_json(ext_obj):
    """Convert bounds to a shapely geojson like spatial object.
    This format is what shapely uses. The output object can be used
    to crop a raster image.

    Parameters
    ----------
    ext_obj: list or geopandas geodataframe
        If provided with a geopandas geodataframe, the extent
        will be generated from that. Otherwise, extent values
        should be in the order: minx, miny, maxx, maxy.

    Returns
    ------
    extent_json: A GeoJSON style dictionary of corner coordinates
    for the extent
```

Reviewer 2: I know this didn't change in the review but I would also update the following

```
Notice that this function's 
docstring includes all parameter inputs and outputs. The description of the function 
is short and to the point (2 to 3 sentences). And the return of the function is 
specified. 
```

to:

```
Notice that this function's 
docstring includes all parameters and return values. The description of the function 
is short and to the point (2 to 3 sentences). And the return of the function is 
specified. 
```

Author: no this is from earthpy. but yes i just copied it over so its likely that we made a mistake there or i made a mistake copying! the lack of spaces is definitely from me here. we gave geopandas some of our code but not this guy. before we knew about rioxarray and all of it's magic powers.

Author: ok fixed! combo of leah-error and maybe leah-error like 5 years ago with the returns but normally i catch that!  
what's really interesting is that it's wrong in the code

https://github.com/earthlab/earthpy/blob/main/earthpy/spatial.py

but the docs print fine - they print _returns_. (not return) 🤷‍♀️

```

```{figure} ../images/sphinx-rendering-extent-to-json-earthpy.png
---
name: directive-fig
width: 80%
---

> Using the above numpy-style docstring in sphinx, the autodoc extension will 

Reviewer 2: ```suggestion
This image shows how the sphinx autodoc extension converts the example docstring above in NumPy format into an entry in the API documentation for the `extent_to_json` function. The output of ...
```

I would rewrite the image caption this way to make explicit the link to the example above and to explain this is API documentation
